src/h3/mod.rs

```diff
@@ -1125,42 +1153,73 @@ impl Connection {
-                let flow_id = b.get_varint()?;
-                return Ok((flow_id, Event::Datagram));
-            },
+        if self.dgram_event_count < self.dgram_poll_threshold &&
+            conn.dgram_recv_front_len().is_some()
+        {
+            match conn.dgram_recv_peek(&mut d, 8) {
+                Ok(_) => {
+                    let mut b = octets::Octets::with_slice(&d);
+                    let flow_id = b.get_varint()?;
 
-            Err(crate::Error::Done) => (),
+                    self.dgram_event_count += 1;
+                    return Ok((flow_id, Event::Datagram));
+                },
 
-            Err(e) => return Err(Error::TransportError(e)),
-        };
+                Err(crate::Error::Done) => (),
+
+                Err(e) => return Err(Error::TransportError(e)),
+            };
+        }
 
-        // Process HTTP/3 data from readable streams.
         for s in conn.readable() {
-            trace!("{} stream id {} is readable", conn.trace_id(), s);
+            if self.stream_event_count < self.stream_poll_threshold ||
+                conn.dgram_recv_front_len().is_none()
+            {
+                trace!("{} stream id {} is readable", conn.trace_id(), s);
 
-            let ev = match self.process_readable_stream(conn, s) {
-                Ok(v) => Some(v),
+                let ev = match self.process_readable_stream(conn, s) {
+                    Ok(v) => Some(v),
 
-                Err(Error::Done) => None,
+                    Err(Error::Done) => None,
 
-                Err(e) => return Err(e),
-            };
+                    Err(e) => return Err(e),
+                };
 
-            if conn.stream_finished(s) {
-                self.finished_streams.push_back(s);
+                if conn.stream_finished(s) {
+                    self.finished_streams.push_back(s);
+                }
+
+                // TODO: check if stream is completed so it can be freed
+                if let Some(ev) = ev {
+                    self.stream_event_count += 1;
+                    return Ok(ev);
+                }
             }
+        }
 
-            // TODO: check if stream is completed so it can be freed
+        // no event generated at this point, so try dgram again
+        if conn.dgram_recv_front_len().is_some() {
+            match conn.dgram_recv_peek(&mut d, 8) {
+                Ok(_) => {
+                    let mut b = octets::Octets::with_slice(&d);
+                    let flow_id = b.get_varint()?;
 
-            if let Some(ev) = ev {
-                return Ok(ev);
-            }
+                    self.dgram_event_count += 1;
+                    return Ok((flow_id, Event::Datagram));
+                },
+
+                Err(crate::Error::Done) => (),
+
+                Err(e) => return Err(Error::TransportError(e)),
+            };
         }
 
         // Process finished streams list once again, to make sure `Finished`
```

Comment:
I _think_ this should go before the second DATAGRAM block, as finished events are prioritized over dgrams in the previous block, but it's probably not super important..

Reply:
done